packages/babel-parser/src/parser/expression.ts

```diff
@@ -273,7 +273,10 @@ export default abstract class ExpressionParser extends LValParser {
   ): N.Expression {
     const startLoc = this.state.startLoc;
     if (this.isContextual(tt._yield)) {
-      if (this.prodParam.hasYield) {
+      if (
+        this.optionFlags & OptionFlags.AllowYieldOutsideFunction ||
```

Comment:
It should be chained with the `!this.scope.inFunction` check, after all we don't allow `yield` within a plain function.